### DISCUSSION TITLE
support socket activation

### DISCUSSION THREAD

**Discussion by eriksjolund** [libuv/libuv#3549] [General]
Add support for [socket activation](https://www.freedesktop.org/software/systemd/man/systemd.socket.html).

__systemd__ can be configured to create a listening TCP socket. When a client connects to this listening TCP socket, __systemd__ will start an application with the socket passed to the application as an open file descriptor descriptor. Make it possible to use such a socket file descriptor in libuv, (e.g. a file descriptor for a TCP socket that is already a listening socket).

### Rationale

* Container runtime support for _socket activation_ has improved lately. [Podman](https://github.com/containers/podman/) version 3.4.0 (released September 2021) supports running OCI containers (i.e. Docker containers) with _socket activation_.  If you have a containerized server that takes requests from the network via a listening TCP port, adding socket activation support opens up the possibility to run the container with the __--network=none__ command-line option. This will restrict the container from accessing the network except via the TCP socket it has been passed by systemd via Podman. Outgoing connections would not be allowed. Running the container with --network=none improves security.

* It would be possible to write a libuv server application that would not be running at all before the first client connects. Less compute resources would be used.

* Having support for socket activation has the additional advantage that template systemd units can be used to create instances of the server with minimal configuration changes. Adding an additional instance of a TCP socket daemon would only require one extra file, a so called systemd drop-in configuration file. For instance https://github.com/eriksjolund/mariadb-podman-socket-activation/blob/main/mariadb-tcp@demo.socket.d/override.conf
    ```
    [Socket]
    ListenStream=127.0.0.1:8090
    ```
    When using UNIX sockets, it's even less preparations to do. No systemd drop-in file is needed as there is no TCP port number to specify. The file path for the UNIX socket could instead be computed by specifying the path with systemd specifier `%i` that will be expanded to the instance name (i.e. the text string between `@` and `.socket` when you activate the socket with `systemctl --user start servicename@instancename.socket`).

**Comment by bnoordhuis** [comment]
Already supported (see https://docs.libuv.org/en/v1.x/tcp.html#c.uv_tcp_open) so converted to a discussion.
